flycheck.el

```diff
@@ -7941,6 +7989,23 @@ Requires Go 1.6 or newer. See URL
   ((warning line-start (file-name) ":" line ":" column ": " (message) line-end))
   :modes go-mode)
 
+(flycheck-define-checker go-staticcheck
+  "A Go checker that performs static analysis and linting using
+the `staticcheck' command. `staticcheck' is the successor to
+`megacheck'; while the latter isn't fully deprecated yet, it's
+recommended to migrate to `staticcheck'.
+
+`staticcheck' is explicitly fully compatible with \"the last two
+versions of go\". `staticheck' can target earlier versions (with
+limited features) if `flycheck-go-version' is set. See URL
+`https://staticcheck.io/'."
+  :command ("staticcheck" "-f" "json"
+            (option-list "-tags" flycheck-go-build-tags concat)
+            (option "-go" flycheck-go-version))
```

Comment:
Since we are not providing a file to check using `source` or its variants, `staticcheck` will be called with the directory of the current file as working directory.  Is this the correct behavior?  I don't know how Go projects are usually structured, but if you have:

```
├── main.go
└── module
    └── foo.go
```
And you open `foo.go`, does the checker still work as expected?  Or should `staticcheck` always run in the top-level directory of your project?  If the latter, then we should pass the `:working-directory` property to the checker.

Reply:
I think the most correct behavior is the current one -- which I will call "current-directory". I can see arguments for `source-inplace`, but I think it leads to usability breakdowns. @fmdkdd, you might not care about all the gory details, but I want to spell them out anyways if anyone ever reads this PR and wants to know what's up. So, here's a tl;dr and also a Long Version.

### tl;dr
Go allows many files to be part of the same package, and it doesn't consider those files "different" -- it only sees the package itself. Using `source-inplace` causes `staticcheck` to be unable to "see" declarations in other files in the same package, causing spurrious "undeclared name" errors.

While we would get correct check results by running in the top-level directory of the project, Go very commonly uses vendoring as its dependency management approach. It's easy to clear a million, two million lines of dependencies. This bogs the checker down hellatiously.

### The full story

For the sake of documenting all of this, here's my reasoning:

A useful thing to know about Go is this: most typically, a directory defines a package, and every file within that directory is part of that package. For example: 

```
├── pkg
│  └── lib
│     ├── boop.go
│     └── lib.go
```

Both `lib.go` and `boop.go` declare themselves as being part of package `lib`. Even in totally "flat" Go projects (i.e. all files at the root level, see [Logrus](https://github.com/sirupsen/logrus) as a good example), all Go files are part of the same package (or else, `package_test`, a special case of the same thing). Importantly: files within the same package can use symbols declared by that package, without declaring an import, _no matter which file they're declared in_. Put another way: Go cares about packages, not specific files. Files are there to allow humans to separate a package in to logical hunks, but Go only sees the package. 

Current-directory and `source-inplace` will both report errors on any Go code directly imported by the file-or-package currently under check; this is good. However: `source-inplace` only checks symbols in the current file, cutting off access to every other file in the package. While this _does_ mean that the checker might report a great number of errors in files outside the one currently under check, I think for Go that behavior is the most correct one.

Reply:
Thansk for the thorough answer.  The "current-directory" behavior looks good to me.